doc/design/request_response_communication.md

```diff
@@ -0,0 +1,118 @@
+
+Since the `Response` is tied to a specific `Request`, the `loan` method takes a `Request` to populate the `Response` with the correct settings.
+Alternatively, a `Request` could have a `createResponse` method which returns a `Response` with the correct settings.
+
+#### Untyped API
+
+![untyped API](diagrams/request_response/untyped_api.svg)
+
+Similar to the the typed API, `loan` takes a pointer to a `RequestHeader` to populate the `ResponseHeader` with the correct settings.
+
+#### Client Port
+
+![client port](diagrams/request_response/client_port.svg)
+
+The `ClientPortData` is located in the shared memory and contain only the data but no methods to access them.
+`ClientPortUser` is the class providing the methods for the user access and `ClientPortRouDi` provides the
+interface RouDi needs to connect the client to the server and to cleanup the port resources.
+
+#### Server Port
+
+![server port](diagrams/request_response/server_port.svg)
+
+Similar to the Client Port, the Server Port has `ServerPortData` which is located in the shared memory and contain only the data but no methods to access them.
+`ServerPortUser` is the class providing the methods for the user access and `ServerPortRouDi` provides the
+interface RouDi needs to connect the client to the server once the server offers its service and to cleanup the port resources.
+
+It must be ensured that only one server with a given `ServiceDescription` can run at a time.
+
+#### Request/Response Header
+
+![rpc header](diagrams/request_response/request_response_header.svg)
+
+Since request and response need to encode different meta-information, we also need different header for the messages.
+The common data is aggregated in `RpcBaseHeader` which contains a `UniquePortId` to the port owning the `ClientChunkQueueData_t` and a sequence ID.
+The `UniquePortId` is used to identify the queue which receives the response and used as identifier for `ChunkDistributor::deliverToPort`.
+This method will iterate over all stored queues and matches the `ChunkQueueData::m_uniqueIdOfOwner`.
+As optimization, the `ChunkDistributor::deliverToPort` returns the index of the queue to do a fast lookup and downgrades to a full lookup if the port IDs don't match.
+```cpp
+class ChunkDistributor {
+    ...
+    cxx::expected<uint64_t, Error> deliverToPort(UniquePortId uniquePortId, uint64_t lastKnownQueueIndex);
+    ...
+};
+```
+
+The sequence ID is used to match a response to a specific request if multiple requests are invoke asynchronously.
+Depending on the client and server options, a request might be dropped or a server could have a worker pool
+which results in sending the responses in a different order than the request were received.
+The sequence ID must be set by the user and also checked by the user on response.
+
+The `RequestHeader` has also the option to specify a message as fire and forget, which means it won't get a response to this request.
```

Comment:
What would be a typical use case for this? And why would generating a response in this use case be cumbersome or unhandy?

Reply:
Sending a status like keep alive or similar where the client does not care about a response

Reply:
And why not use a publisher for that? A fire and forget request seems to me a typical publisher use case.

Reply:
Because there can be many subscriber for one publisher but only one server for a client. Furthermore, we need this to support bridges to other technologies who also have fire and forget RPCs